Vendor updated version of docker/distribution

aaronlehmann opened the pull request:
This updates the vendored docker/distribution to the current master
branch.

Note the following changes:
- The manifest package was split into manifest/schema1. Most references
  to the manifest package in the engine needed to be updated to use
  schema1 instead.
- Validation functions in api/v2 were replaced by the
  distribution/reference package. The engine code has been updated to
  use the reference package for validation where necessary. A future PR
  will change the engine to use the types defined in
  distribution/reference more comprehensively.
- The reference package explicitly allows double _ characters in
  repository names. registry_test.go was updated for this.
- TestPullFailsWithAlteredManifest was corrupting the manifest JSON, now
  that the schema1 package unmarshals the correct payload. The test is
  being changed to modify the JSON without affecting its length, which
  allows the pull to succeed to the point where digest validation
  happens.

cc @docker/distribution-maintainers @tonistiigi

Commit:
Vendor updated version of docker/distribution

This updates the vendored docker/distribution to the current master
branch.

Note the following changes:

- The manifest package was split into manifest/schema1. Most references
  to the manifest package in the engine needed to be updated to use
  schema1 instead.

- Validation functions in api/v2 were replaced by the
  distribution/reference package. The engine code has been updated to
  use the reference package for validation where necessary. A future PR
  will change the engine to use the types defined in
  distribution/reference more comprehensively.

- The reference package explicitly allows double _ characters in
  repository names. registry_test.go was updated for this.

- TestPullFailsWithAlteredManifest was corrupting the manifest JSON, now
  that the schema1 package unmarshals the correct payload. The test is
  being changed to modify the JSON without affecting its length, which
  allows the pull to succeed to the point where digest validation
  happens.

Signed-off-by: Aaron Lehmann <aaron.lehmann@docker.com>

dmcgowan commented on registry/registry_test.go:
```diff
@@ -803,8 +803,7 @@ func TestValidRemoteName(t *testing.T) {
 		"_docker/_docker",
 
-		// Disallow consecutive underscores and periods.
-		"dock__er/docker",
```
My preference with these cases is to move from invalid to valid to make the behavior change more explicit

aaronlehmann replied:
Updated to move it to valid.

stevvooe commented on graph/pull_v2_test.go:
```diff
@@ -167,28 +167,4 @@ func TestValidateManifest(t *testing.T) {
-
-	// Manifest with no signature
-
-	expectedWholeFileDigest := "7ec3615a120efcdfc270e9c7ea4183330775a3e52a09e2efb194b9a7c18e5ff7"
```
I am not sure about removing this. Was it failing?

aaronlehmann replied:
Yes. The Unmarshal call a few lines below fails because the current `SignedManifest` unmarshal method uses the signature section as part of the unmarshaling process. If the signature is missing, JSON unmarshaling fails.

stevvooe replied:
That is the definition of a signed manifest. If there is no signature, it is not valid. Rather than removing the test, can we change it to make it work correctly?

aaronlehmann replied:
This particular test is for the `validateManifest` function. It's no longer possible to deserialize the manifest in order to call `validateManifest`. We could have a separate test to ensure that unmarshaling fails. I think such a test would belong under `distribution/manifest/schema1`. Would you like a PR to distribution to add a test there?

stevvooe replied:
In the current world, it might be a bug that one cannot unmarshal into an SignedManifest.

And yes, this is completely against what I said above. Let's make sure that we are at least testing `validateManifest`.

aaronlehmann commented:
Updated the commit to vendor a slightly newer version of distribution, bringing in a small tweak to the reference package that simplifies the usage of `WithTag`.

Also changed the commit to move the `dock__er/docker` test case in `TestValidRemoteName` from invalid to valid instead of removing it, and to preserve the "no signature" case of `TestValidateManifest`.

dmcgowan commented:
LGTM

stevvooe commented:
LGTM

tonistiigi commented:
LGTM
